### PR TITLE
feat(api): run step filters - is online logic

### PR DESCRIPTION
**Pull request by LetItRock**
### What change does this PR introduce?

Implemented the step filters logic for the "Is Online Filters".

### Why was this change needed?

This is a part of the "Is Online Filters" functionality.

### CODE REVIEW

File: apps/api/src/app/events/usecases/trigger-event/message-matcher.service.spec.ts [hunk 1]
```diff
@@ -557,6 +558,189 @@ describe('Message filter matcher', function () {
...
+      lastOnlineAt: subDuration ? sub(new Date(), subDuration).toISOString() : new Date().toISOString(),
+    });
+
+    describe('isOnline', () => {
+      it('allows to process if the subscriber is online', async () => {
+        const matcher = new MessageMatcher(
+          { findOne: () => Promise.resolve(getSubscriber()) } as any,
+          undefined as any,
+          undefined as any
+        );
```
Comment (LetItRock) [inline apps/api/src/app/events/usecases/trigger-event/message-matcher.service.spec.ts:579]: I couldn't find a better way how to pass a mocked `SubscriberRepository` instance with Sinon to the `MessageMatcher`, so decided to do that this way.

File: libs/dal/src/repositories/subscriber/subscriber.schema.ts [hunk 2]
```diff
@@ -27,7 +27,7 @@ const subscriberSchema = new Schema(
       type: Schema.Types.Boolean,
       default: false,
     },
-    lastOnlineAt: Schema.Types.String,
+    lastOnlineAt: Schema.Types.Date,
```
Comment (LetItRock) [inline libs/dal/src/repositories/subscriber/subscriber.schema.ts:30]: in DB we do store it as a date

File: apps/api/src/app/events/usecases/trigger-event/message-matcher.service.ts [hunk 3]
```diff
@@ -112,6 +132,28 @@ export class MessageMatcher {
...
+    const subscriber = await this.subscriberRepository.findOne({
+      _id: command.subscriberId,
+      _organizationId: command.organizationId,
+      _environmentId: command.environmentId,
+    });
+
+    const isOnlineMatch = subscriber?.isOnline === filter.value;
+    if (filter.on === 'isOnline') {
+      return isOnlineMatch;
+    }
```
Comment (LetItRock) [inline apps/api/src/app/events/usecases/trigger-event/message-matcher.service.ts:148]: for the `isOnline` filter we just do check if the `isOnline` value on the subscriber matches with the filter value

File: apps/api/src/app/events/usecases/trigger-event/message-matcher.service.ts [hunk 4]
```diff
@@ -112,6 +132,28 @@ export class MessageMatcher {
...
+
+    const currentDate = new Date();
+    const lastOnlineAt = subscriber?.lastOnlineAt ? parseISO(subscriber?.lastOnlineAt) : new Date();
+    const diff = differenceIn(currentDate, lastOnlineAt, filter.timeOperator);
+
+    return subscriber?.isOnline || (!subscriber?.isOnline && diff <= filter.value);
```
Comment (LetItRock) [inline apps/api/src/app/events/usecases/trigger-event/message-matcher.service.ts:154]: for the `isOnlineInLast` we do check:
- if the subscriber is online
- or if the subscriber is offline but was online in the last X minutes, hours, days

**Review by BiswaViraj — APPROVED** [review]
🚀

**Pull request merged**